Realm/ObjectStore/object_store.hpp

```diff
@@ -0,0 +1,112 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2014 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OBJECT_STORE_HPP
+#define REALM_OBJECT_STORE_HPP
+
+#include <map>
+#include <vector>
+#include <functional>
+
+#include "object_schema.hpp"
+#include "object_store_exceptions.hpp"
+
+namespace realm {
+    class Group;
+    class StringData;
+    class Table;
+    template<typename T> class BasicTableRef;
+    typedef BasicTableRef<Table> TableRef;
+
+    class ObjectStore {
+    public:
+        // Schema version used for uninitialized Realms
+        static const uint64_t NotVersioned;
+
+        // get the last set schema version
+        static uint64_t get_schema_version(Group *group);
+
+        // checks if a migration is required for a given schema version
+        static bool is_migration_required(realm::Group *group, uint64_t new_version);
+
+        // verify a target schema against its table, setting the table_column property on each schema object
+        // updates the column mapping on the target_schema
+        // if no table is provided it is fetched from the group
+        // returns array of validation errors
+        static std::vector<std::string> validate_schema(Group *group, ObjectSchema &target_schema, Table *cached_table = nullptr);
```

Comment:
Yes. To avoid merge hell we want to minimize the amount of time that we have ported and unmerged versions of the current obj-c code.